app/src/main/java/com/example/chattutorial/ChannelActivity3.kt

```diff
@@ -45,35 +43,21 @@ class ChannelActivity3 : AppCompatActivity(R.layout.activity_channel_3) {
                 }
             }
 
-        val channelController = ChatClient.instance().channel(cid)
-        val currentlyTyping = MutableLiveData<Set<String>>(emptySet())
-
-        channelController.subscribe {
-            val typing = currentlyTyping.value ?: emptySet()
-            val typingCopy: MutableSet<String> = typing.toMutableSet()
-            when (it) {
-                is TypingStartEvent -> {
-                    val name = it.user.extraData["name"] as String
-                    typingCopy += name
-                    currentlyTyping.postValue(typingCopy)
+        channelHeaderView.text = "nobody is typing"
+        val currentlyTyping = mutableSetOf<String>()
+        ChatClient.instance().channel(cid).subscribe { event ->
+            when (event) {
+                is TypingStartEvent -> event.user.extraData["name"]?.let {
+                    currentlyTyping.add(it as String)
+                    updateChannelHeaderView(currentlyTyping)
                 }
-                is TypingStopEvent -> {
-                    val name = it.user.extraData["name"] as String
-                    typingCopy -= name
-                    currentlyTyping.postValue(typingCopy)
+                is TypingStopEvent -> event.user.extraData["name"]?.let {
+                    currentlyTyping.remove(it as String)
+                    updateChannelHeaderView(currentlyTyping)
                 }
             }
         }
 
-        val typingObserver = Observer<Set<String>> { users ->
-            var typing = "nobody is typing"
-            if (users.isNotEmpty()) {
-                typing = "typing: " + users.joinToString(", ")
-            }
-            channelHeaderView.text = typing
-        }
-        currentlyTyping.observe(this, typingObserver)
```

Comment:
This observation through `LiveData` is lifecycle-safe, it makes sure that updates only happen to the header View when `this` lifecycle is in a _started_ state. Are we purposefully making this unsafe in the sample? What if clients copy this code directly into their production applications?

Reply:
yeah you're right. Looks simpler but it's unsafe to do that in sample. I'll go back to the old implementation